### PR TITLE
graphdriver/btrfs: needs kernel headers >= 4.12, not >= 4.7

### PR DESCRIPTION
**Pull request by AkihiroSuda**
**- What I did**

Follow-up to
- https://github.com/moby/moby/pull/44761

Updated `project/PACKAGERS.md` to clarify that the minimum requirement is 4.12, not 4.7.

The 4.7 kernel tree has https://github.com/torvalds/linux/blob/v4.7/include/uapi/linux/btrfs_tree.h but it was not installed to `/usr/include/linux/btrfs_tree.h` until kernel 4.12: https://github.com/torvalds/linux/commit/fcc8487d477a3452a1d0ccbdd4c5e0e1e3cb8bed

Also added a `LINUX_VERSION_CODE` check in `daemon/graphdriver/btrfs/btrfs.go`

**- How to verify it**

Build this Dockerfile with `LINUX_VERSION=4.12` and `LINUX_VERSION=4.7`. 
The latter one fails with `linux/btrfs_tree.h: No such file or directory`.

```dockerfile
ARG GOLANG_VERSION=1.19
ARG LINUX_VERSION=4.12
# 2023-01-07
ARG MOBY_VERSION=6834304febca3c0610d1061bd9612ee99020df34

FROM golang:${GOLANG_VERSION}
ARG LINUX_VERSION
RUN curl -sSL -O "https://mirrors.edge.kernel.org/pub/linux/kernel/v$(echo ${LINUX_VERSION} | cut -d. -f1).x/linux-${LINUX_VERSION}.tar.gz"
RUN tar Cxzf / "linux-${LINUX_VERSION}.tar.gz" && \
  cd "/linux-${LINUX_VERSION}" && \
  make headers_install INSTALL_HDR_PATH=/usr2 && \
  for f in /usr2/include/*; do rm -rf "/usr/include/$(basename ${f})"; done && \
  cp -a /usr2/include /usr

RUN mkdir -p /go/src/github.com/docker && \
  git clone https://github.com/moby/moby.git /go/src/github.com/docker/docker
WORKDIR /go/src/github.com/docker/docker
ARG MOBY_VERSION
RUN git checkout "${MOBY_VERSION}"
ENV GO111MODULE=off
RUN go build ./daemon/graphdriver/btrfs
```

**- A picture of a cute animal (not mandatory but encouraged)**
🐧

### CODE REVIEW

File: daemon/graphdriver/btrfs/btrfs.go [hunk 1]
```diff
@@ -7,6 +7,11 @@ package btrfs // import "github.com/docker/docker/daemon/graphdriver/btrfs"
 #include <stdlib.h>
 #include <stdio.h>
 #include <dirent.h>
+#include <linux/version.h>
+#if LINUX_VERSION_CODE < KERNEL_VERSION(4,12,0)
+#error "Headers from kernel >= 4.12 are required on compilation time (not on run time)"
+#error "Hint: set 'DOCKER_BUILDTAGS=exclude_graphdriver_btrfs' to exclude btrfs"
+#endif
```
Comment (neersighted) [inline daemon/graphdriver/btrfs/btrfs.go:14]:
```suggestion

#include <linux/version.h>
#if LINUX_VERSION_CODE < KERNEL_VERSION(4,12,0)
    #error "Headers from kernel >= 4.12 are required to build with Btrfs support."
    #error "HINT: Set 'DOCKER_BUILDTAGS=exclude_graphdriver_btrfs' build without Btrfs."
#endif
```

**Review by neersighted — CHANGES_REQUESTED** [review]
LGTM with slight formatting nits.

I've verified that the commit that properly exports this header was not backported to any LTS version in gregkh's tree: https://github.com/gregkh/linux/commit/fcc8487d477a3452a1d0ccbdd4c5e0e1e3cb8bed

It looks like this snuck through the first time as I was bind-mounting in a kernel tree to containers and forgot to ensure I was at the right tag; my bad there.

I've also confirmed that this should not change assumptions/statements about distribution support:

* [RHEL 8](https://access.redhat.com/articles/3078) is on 4.18
* [Ubuntu 16.04](https://ubuntu.com/about/release-cycle#ubuntu-kernel-release-cycle) is on 4.15
* [SLES 12 SP4](https://www.suse.com/support/kb/doc/?id=000019587) is on 4.12
* [Debian oldstable](https://tracker.debian.org/pkg/linux) is on 4.19

File: daemon/graphdriver/btrfs/btrfs.go (hunk 1 above)
Comment (AkihiroSuda) [reply daemon/graphdriver/btrfs/btrfs.go:14]: done

Comment (neersighted) [reply daemon/graphdriver/btrfs/btrfs.go]: Looks like I managed to drop a word -- `to build without Btrfs.`

Also, could we add the line break between the libc and kernel version header?

Comment (AkihiroSuda) [reply daemon/graphdriver/btrfs/btrfs.go]: done

**Review by neersighted — APPROVED** [review]

**Pull request merged**